Structured Qualifications Clean Up

raamSoftwire opened the pull request:
## Context

While reviewing the structured qualification work, quite a few bugs were found. This card deals with cleaning up all of these bugs

## Changes proposed in this pull request

Checklist documented on ticket

#1 Updated English GCSE error messages 

![image](https://user-images.githubusercontent.com/33458055/100471289-e698be00-30d1-11eb-9e20-cb2a3b42902a.png)

![image](https://user-images.githubusercontent.com/33458055/100594801-0c9ca900-32f2-11eb-8a9a-5a9e842f9987.png)

![image](https://user-images.githubusercontent.com/33458055/100628420-189f5f80-3320-11eb-8d5d-5993f249baf4.png)

![image](https://user-images.githubusercontent.com/33458055/100870162-0e9f6d00-3496-11eb-93d7-98540d20e148.png)

![image](https://user-images.githubusercontent.com/33458055/100870190-1ced8900-3496-11eb-9102-564afffaaf86.png)


## Guidance to review

<!-- How could someone else check this work? Which parts do you want more feedback on? -->

## Link to Trello card

https://trello.com/c/edEWlzAL/2599-dev-structured-qualifications-clean-up-work

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

davidgisbey commented on app/models/application_qualification.rb:
```diff
@@ -74,7 +74,7 @@ def incomplete_other_qualification?
       end
     else
-      return true if EXPECTED_OTHER_QUALIFICATION_DATA.any? do |field_name|
+      return true if EXPECTED_OTHER_QUALIFICATION_DATA.reject { |k| k == :grade }.any? do |field_name|
```
Can we just remove `:grade` from the constant rather than reject it?

davidgisbey commented:
Looks good, just found this while testing locally.

![image](https://user-images.githubusercontent.com/42515961/100871958-f8df7700-3498-11eb-81d1-9bf7f648a8e0.png)

davidgisbey commented:
One more. We should be using an autocomplete here. Just validating in the form
 
![image](https://user-images.githubusercontent.com/42515961/100872731-2a0c7700-349a-11eb-82b3-a01bdabf0f14.png)

davidgisbey commented on app/components/gcse_qualification_cards_component.rb:
```diff
@@ -47,14 +53,19 @@ def naric_statement(qualification)
+      ["#{qualification.grade} (Single award)"]
+    when ->(_n) { qualification.structured_grades }
       grades = JSON.parse(qualification.structured_grades)
       grades.map { |k, v,| "#{v} (#{k.humanize.titleize})" }
```
This causes double award to upcase. We should probably have a bespoke bit of logic for when grade.key == english_double_award 

![image](https://user-images.githubusercontent.com/42515961/100874332-51fcda00-349c-11eb-9b45-3b279343ad7f.png)

davidgisbey commented on app/helpers/gcse_qualification_helper.rb:
```diff
@@ -15,4 +15,24 @@ def hint_for_gcse_edit_grade(subject, qualification_type)
   end
+
+  def grade_step_title(subject, qualification_type)
+    subject = subject.capitalize if subject == 'english'
```
Please can we also capitalise Science and Maths here so it's consistent across the subjects

TobyRet replied:
@paulrobertlloyd ☝️ - can you confirm? Thought it was only English that was supposed to be capitalised?

davidgisbey replied:
You're probably right. My grasp of the English language is tenuous at best.

paulrobertlloyd replied:
Yep, only English is capitalised.

davidgisbey reviewed (changes requested):
Really great work Raam, I've found a few things while manual testing.

TobyRet commented on app/components/candidate_interface/gcse_qualification_review_component.rb:
```diff
@@ -65,14 +65,19 @@ def grade_row
+        "#{application_qualification.grade} (Double award)"
+      when ApplicationQualification::SCIENCE_SINGLE_AWARD
+        "#{application_qualification.grade} (Single award)"
+      when ->(subject) { subject == 'english' && application_qualification.structured_grades }
```
Interesting, did not know you could use lambdas in case statements 👍

raamSoftwire replied:
The reason I needed this was because it was the only way I could find to allow an AND statement. I don't think it's particularly readable, so please shout if there's a more standard way.

davidgisbey replied:
+1

TobyRet commented on app/models/application_qualification.rb:
```diff
@@ -74,7 +74,7 @@ def incomplete_other_qualification?
       end
     else
-      return true if EXPECTED_OTHER_QUALIFICATION_DATA.any? do |field_name|
+      return true if EXPECTED_OTHER_QUALIFICATION_DATA.reject { |k| k == :grade }.any? do |field_name|
```
I'm wondering the entire `incomplete_other_qualification` could be re-written to make it more readable or at the min, extract the if clauses to methods for readability?

e.g.

```
def incomplete_other_qualification?
   (qualification_type == 'non_uk' && incomplete_international_data?) || incomplete_other_data?
end

```

raamSoftwire replied:
This was a great suggestion Toby, thanks! I worked it through step by step and ended up at the same solution. This method is now much, much easier to read.

raamSoftwire replied:
This actually broke some tests and led me to deep a bit deeper. For non UK qualifications `incomplete_other_data?` is true and the way the old way worked was by returning early so it didn't get that far. 

```  def incomplete_other_qualification?
    if qualification_type == 'non_uk'
      incomplete_international_data?
    elsif other?
      incomplete_other_qualification_data?
    else
      false
    end
  end

TobyRet commented on config/locales/application_form.yml:
```diff
@@ -843,26 +843,26 @@ en:
             grade_english_single:
-              blank: Enter your grade
-              invalid: Enter a real grade
+              blank: Enter your English (Single award) grade
```
Probably need to ensure consistency with science validation messages. I think @paulrobertlloyd wanted ‘science’ dropped from each as this was deemed unnecessary.

Science validation messages - https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/config/locales/application_form.yml#L811

raamSoftwire replied:
I wonder if English GCSEs are a bit different? 

I don't think people reference to Literature or Language GCSEs rather than English Literature or English Language. I'd be keen to see what Paul thinks here.

raamSoftwire replied:
There is also English single and double as well as English Studies single and double, so I think it'd be clearer to leave the extra clarification in here.

TobyRet reviewed (commented):
Looking good - some comments added

raamSoftwire commented:
![image](https://user-images.githubusercontent.com/33458055/100993930-eb37f900-354d-11eb-8dab-b796d15dabae.png)

I've now updated the old English page to bring it more in line with how the newer page works 
- there is no autocomplete
- I've updated the hint to include double awards because we already accept them, but don't mention it
- made the width the right size for double award grades

davidgisbey commented on app/views/candidate_interface/other_qualifications/details/_form.html.erb:
```diff
@@ -7,7 +7,7 @@
     <%= tag.div(id: 'subject-autocomplete-data', data: { source: subject_autocomplete_data }) if subject_autocomplete_data %>
     <%= f.govuk_text_field :grade,
-                           label: { text: t('application_form.other_qualification.grade.optional_label'), size: 'm' },
+                           label: { text: @form.qualification_type === "Other" ? t('application_form.other_qualification.grade.optional_label') : t('application_form.other_qualification.grade.label'), size: 'm' },
```
== 'Other'

davidgisbey reviewed (approved):
Great work 👍
